File: src/types/service.ts
/**
 * Service system type definitions
 *
 * Types for the modular service automation system.
 * Matches Rust types in src-tauri/src/types/service.rs
 */

// =============================================================================
// Service Identification
// =============================================================================

/** Unique identifier for a service */
export type ServiceId = string;

/** Unique identifier for a service run/report */
export type ReportId = string;

// =============================================================================
// Service Definitions
// =============================================================================

/** Option for select-type service options */
export interface SelectOption {
  value: string;
  label: string;
}

/** Schema for a service option field */
export interface ServiceOptionSchema {
  /** Option identifier */
  id: string;
  /** Display label */
  label: string;
  /** Option type: "number", "string", "boolean", "select" */
  optionType: 'number' | 'string' | 'boolean' | 'select';
  /** Default value */
  defaultValue: unknown;
  /** For number type: minimum value */
  min?: number;
  /** For number type: maximum value */
  max?: number;
  /** For select type: available options */
  options?: SelectOption[];
  /** Help text for the option */
  description?: string;
}

/** Definition of a service that can be run */
export interface ServiceDefinition {
  /** Unique service identifier */
  id: ServiceId;
  /** Display name */
  name: string;
  /** Description of what the service does */
  description: string;
  /** Category for grouping (e.g., "diagnostics", "cleanup", "security") */
  category: string;
  /** Estimated duration in seconds (for progress estimation) */
  estimatedDurationSecs: number;
  /** Program IDs required to run this service (from programs.json) */
  requiredPrograms: string[];
  /** Configurable options for this service */
  options: ServiceOptionSchema[];
  /** Icon name (lucide icon identifier) */
  icon: string;
}

// =============================================================================
// Service Presets
// =============================================================================

/** Service configuration within a preset */
export interface PresetServiceConfig {
  /** Service ID */
  serviceId: ServiceId;
  /** Whether enabled by default in this preset */
  enabled: boolean;
  /** Default options for this service in the preset */
  options: Record<string, unknown>;
}

/** A preset configuration of services */
export interface ServicePreset {
  /** Preset identifier */
  id: string;
  /** Display name */
  name: string;
  /** Description */
  description: string;
  /** Services included in this preset (with default options) */
  services: PresetServiceConfig[];
  /** Icon name */
  icon: string;
  /** Accent color for the card */
  color: string;
}

// =============================================================================
// Service Queue
// =============================================================================

/** An item in the service run queue */
export interface ServiceQueueItem {
  /** Service ID */
  serviceId: ServiceId;
  /** Whether this service is enabled for the run */
  enabled: boolean;
  /** Order in the queue (0-indexed) */
  order: number;
  /** User-configured options */
  options: Record<string, unknown>;
}

// =============================================================================
// Service Results
// =============================================================================

/** Severity level for a finding */
export type FindingSeverity = 'info' | 'success' | 'warning' | 'error' | 'critical';

/** A single finding from a service */
export interface ServiceFinding {
  /** Severity level */
  severity: FindingSeverity;
  /** Short title */
  title: string;
  /** Detailed description */
  description: string;
  /** Recommended action (if any) */
  recommendation?: string;
  /** Raw data (for technical details) */
  data?: unknown;
}

/** Result of running a single service */
export interface ServiceResult {
  /** Service ID that was run */
  serviceId: ServiceId;
  /** Whether the service completed successfully */
  success: boolean;
  /** Error message if failed */
  error?: string;
  /** Execution time in milliseconds */
  durationMs: number;
  /** Findings from the service */
  findings: ServiceFinding[];
  /** Log output from the service */
  logs: string[];
}

/** Status of a service run */
export type ServiceRunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

/** A complete service run report */
export interface ServiceReport {
  /** Unique report ID */
  id: ReportId;
  /** When the run started (ISO string) */
  startedAt: string;
  /** When the run completed (ISO string, null if still running) */
  completedAt?: string;
  /** Overall status */
  status: ServiceRunStatus;
  /** Total duration in milliseconds */
  totalDurationMs?: number;
  /** Queue that was run */
  queue: ServiceQueueItem[];
  /** Results for each service */
  results: ServiceResult[];
  /** Index of currently running service (for progress) */
  currentServiceIndex?: number;
}

// =============================================================================
// Service Run State
// =============================================================================

/** Global service run state */
export interface ServiceRunState {
  /** Whether a service run is currently active */
  isRunning: boolean;
  /** Current report being generated */
  currentReport?: ServiceReport;
}

// =============================================================================
// UI State
// =============================================================================

/** Current phase of the service page */
export type ServicePhase = 'presets' | 'queue' | 'running' | 'results';

/** Service page state */
export interface ServicePageState {
  /** Current phase */
  phase: ServicePhase;
  /** Selected preset (if any) */
  selectedPreset?: ServicePreset;
  /** Current queue */
  queue: ServiceQueueItem[];
  /** Current or last report */
  report?: ServiceReport;
}

// =============================================================================
// Constants
// =============================================================================

/** Map of severity to display info */
export const SEVERITY_INFO: Record<FindingSeverity, { label: string; color: string }> = {
  info: { label: 'Info', color: 'blue' },
  success: { label: 'Success', color: 'green' },
  warning: { label: 'Warning', color: 'yellow' },
  error: { label: 'Error', color: 'red' },
  critical: { label: 'Critical', color: 'purple' },
};

/** Map of run status to display info */
export const STATUS_INFO: Record<ServiceRunStatus, { label: string; color: string }> = {
  pending: { label: 'Pending', color: 'gray' },
  running: { label: 'Running', color: 'blue' },
  completed: { label: 'Completed', color: 'green' },
  failed: { label: 'Failed', color: 'red' },
  cancelled: { label: 'Cancelled', color: 'yellow' },
};

File: src-tauri/src/types/service.rs
//! Service system type definitions
//!
//! Types for the modular service automation system.
//! Services run diagnostic and maintenance tasks on the system.

use serde::{Deserialize, Serialize};

// =============================================================================
// Service Identification
// =============================================================================

/// Unique identifier for a service
pub type ServiceId = String;

/// Unique identifier for a service run/report
pub type ReportId = String;

// =============================================================================
// Service Definitions
// =============================================================================

/// Schema for a service option field
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceOptionSchema {
    /// Option identifier
    pub id: String,
    /// Display label
    pub label: String,
    /// Option type: "number", "string", "boolean", "select"
    pub option_type: String,
    /// Default value (JSON encoded)
    pub default_value: serde_json::Value,
    /// For number type: minimum value
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<f64>,
    /// For number type: maximum value
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<f64>,
    /// For select type: available options
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<Vec<SelectOption>>,
    /// Help text for the option
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// Option for select-type service options
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelectOption {
    pub value: String,
    pub label: String,
}

/// Definition of a service that can be run
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceDefinition {
    /// Unique service identifier
    pub id: ServiceId,
    /// Display name
    pub name: String,
    /// Description of what the service does
    pub description: String,
    /// Category for grouping (e.g., "diagnostics", "cleanup", "security")
    pub category: String,
    /// Estimated duration in seconds (for progress estimation)
    pub estimated_duration_secs: u32,
    /// Program IDs required to run this service (from programs.json)
    /// Empty vec means no external programs needed
    pub required_programs: Vec<String>,
    /// Configurable options for this service
    pub options: Vec<ServiceOptionSchema>,
    /// Icon name (lucide icon identifier)
    pub icon: String,
}

// =============================================================================
// Service Presets
// =============================================================================

/// A preset configuration of services
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServicePreset {
    /// Preset identifier
    pub id: String,
    /// Display name
    pub name: String,
    /// Description
    pub description: String,
    /// Services included in this preset (with default options)
    pub services: Vec<PresetServiceConfig>,
    /// Icon name
    pub icon: String,
    /// Accent color for the card
    pub color: String,
}

/// Service configuration within a preset
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresetServiceConfig {
    /// Service ID
    pub service_id: ServiceId,
    /// Whether enabled by default in this preset
    pub enabled: bool,
    /// Default options for this service in the preset
    pub options: serde_json::Value,
}

// =============================================================================
// Service Queue
// =============================================================================

/// An item in the service run queue
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceQueueItem {
    /// Service ID
    pub service_id: ServiceId,
    /// Whether this service is enabled for the run
    pub enabled: bool,
    /// Order in the queue (0-indexed)
    pub order: u32,
    /// User-configured options
    pub options: serde_json::Value,
}

// =============================================================================
// Service Results
// =============================================================================

/// Severity level for a finding
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum FindingSeverity {
    Info,
    Success,
    Warning,
    Error,
    Critical,
}

/// A single finding from a service
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceFinding {
    /// Severity level
    pub severity: FindingSeverity,
    /// Short title
    pub title: String,
    /// Detailed description
    pub description: String,
    /// Recommended action (if any)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommendation: Option<String>,
    /// Raw data (for technical details)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

/// Result of running a single service
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceResult {
    /// Service ID that was run
    pub service_id: ServiceId,
    /// Whether the service completed successfully
    pub success: bool,
    /// Error message if failed
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    /// Execution time in milliseconds
    pub duration_ms: u64,
    /// Findings from the service
    pub findings: Vec<ServiceFinding>,
    /// Log output from the service
    pub logs: Vec<String>,
}

/// Status of a service run
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ServiceRunStatus {
    /// Run is queued but not started
    Pending,
    /// Currently running
    Running,
    /// Completed successfully
    Completed,
    /// Failed with error
    Failed,
    /// Cancelled by user
    Cancelled,
}

/// A complete service run report
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceReport {
    /// Unique report ID
    pub id: ReportId,
    /// When the run started (ISO string)
    pub started_at: String,
    /// When the run completed (ISO string, null if still running)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
    /// Overall status
    pub status: ServiceRunStatus,
    /// Total duration in milliseconds
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_duration_ms: Option<u64>,
    /// Queue that was run
    pub queue: Vec<ServiceQueueItem>,
    /// Results for each service (keyed by service_id)
    pub results: Vec<ServiceResult>,
    /// Index of currently running service (for progress)
    pub current_service_index: Option<usize>,
    /// Technician who performed the service (business mode)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub technician_name: Option<String>,
    /// Customer name (business mode)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub customer_name: Option<String>,
}

// =============================================================================
// Service Run State (for persistent running across tab navigation)
// =============================================================================

/// Global service run state
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceRunState {
    /// Whether a service run is currently active
    pub is_running: bool,
    /// Current report being generated
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_report: Option<ServiceReport>,
}

impl Default for ServiceRunState {
    fn default() -> Self {
        Self {
            is_running: false,
            current_report: None,
        }
    }
}
